### PR TITLE
Add ws_proxy option for connecting gateway through a proxy.

### PR DESCRIPTION
**Pull request by zzzzRuby**
#2479

### CODE REVIEW

**Comment by cheesycod** [comment]
This would help me as well. I forked nirn to hackily replace the ``ws`` url.

**Comment by GnomedDev** [comment]
I'm kinda wondering why this is required at all, can't the gateway url just be exposed (if not already) to point to the proxy, instead of having this proxy option, and why is the manual http-over-tcp required? None of this is commented or explained.

**Comment by zzzzRuby** [comment]
> I'm kinda wondering why this is required at all, can't the gateway url just be exposed (if not already) to point to the proxy, instead of having this proxy option, and why is the manual http-over-tcp required? None of this is commented or explained.

Sorry I'm new to opensource.
In some network environments like Chinese networks, there must be a proxy to connect with discord gateway. There're two options to do this, the first one is open a tunnel to the proxy likes this pr does, the second one is creating a custom gateway and redirect all message through a tunnel to the proxy. The second one seems too complex.
As to the manual http-over-tcp, I'm not sure if I can add a new dependency in Cargo.toml.

**Comment by GnomedDev** [comment]
Can you not perform this proxying system wide, instead of coding support for it into every application? I'm honestly not familar with this.

**Comment by zzzzRuby** [comment]
> Can you not perform this proxying system wide, instead of coding support for it into every application? I'm honestly not familar with this.

Normally, libraries like libcurl or reqwest get a default proxy from environment or system. But tungstenite doesn't support this and they seem have no interest in this (snapview/tungstenite-rs#177), the only choice to perform a proxy system wide is proxychains-like-tools. But I don't think proxychains is a good idea, it depends on linking libc dynamically, so on some distro like alpine it's hard to use with rust program and sometimes not work.

File: src/client/mod.rs [hunk 1]
```diff
@@ -293,6 +295,23 @@ impl ClientBuilder {
...
+    pub fn ws_proxy<T: Into<String>>(mut self, proxy: T) -> Self {
+        self.ws_proxy = Some(proxy.into());
+        self
+    }
+
+    /// Remove websocket proxy.
+    pub fn no_ws_proxy(mut self) -> Self {
+        self.ws_proxy = None;
+        self
+    }
```
Comment (GnomedDev) [inline src/client/mod.rs]: Can this method be removed, I don't see a useful point to it.

File: src/client/mod.rs [hunk 2]
```diff
@@ -351,6 +371,7 @@ impl IntoFuture for ClientBuilder {
                 #[cfg(feature = "voice")]
                 voice_manager: voice_manager.as_ref().map(Arc::clone),
                 ws_url: Arc::clone(&ws_url),
+                ws_proxy,
```
Comment (GnomedDev) [inline src/client/mod.rs:374]: Can this be wrapped in `Arc` like `ws_url`, and preprocessed into `Url` to fail here if the user provides an invalid URL.

File: src/gateway/ws.rs [hunk 3]
```diff
@@ -101,13 +108,72 @@ const TIMEOUT: Duration = Duration::from_millis(500);
 const DECOMPRESSION_MULTIPLIER: usize = 3;
 
 impl WsClient {
-    pub(crate) async fn connect(url: Url) -> Result<Self> {
+    async fn connect_with_proxy_async(
```
Comment (GnomedDev) [inline src/gateway/ws.rs:109]: This whole method should be replaced with something using `reqwest`, which is already in the library.

**Review by GnomedDev — COMMENTED** [review]
Okay, I understand now, the situation sucks (this should be on tungstenite) but if we can make a positive impact lets go for this

File: src/gateway/ws.rs (hunk 3 above)
Comment (Itsusinn) [reply src/gateway/ws.rs:109]: > This whole method should be replaced with something using `reqwest`, which is already in the library.

Reqwest doesn't expose API that can be used in websocket proxy. But there is an implementation that borrows from reqwest's design.
https://github.com/snapview/tungstenite-rs/issues/177#issuecomment-1352808621

File: src/client/mod.rs (hunk 1 above)
Comment (zzzzRuby) [reply src/client/mod.rs]: Removed it.

File: src/gateway/ws.rs (hunk 3 above)
Comment (zzzzRuby) [reply src/gateway/ws.rs:109]: Just like [Itsusinn](https://github.com/Itsusinn) said. I replaced the http connect code with async-http-proxy[https://crates.io/crates/async-http-proxy] which is mentioned in that issue.

File: src/client/mod.rs (hunk 2 above)
Comment (zzzzRuby) [reply src/client/mod.rs:374]: Yes it can. Finished

Comment (mkrasnitski) [reply src/client/mod.rs:374]: `ws_url` is only parsed into a `Url` inside of the `gateway::shard::connect` helper function. I think for consistency's sake the same should be done with `ws_proxy` - a followup PR could move the parsing logic for both to be earlier in the initialization stage.

**Comment by mlesin** [comment]
Are there any plans to merge this?

**Comment by GnomedDev** [comment]
As the author seems to be MIA, no. If anyone wants to rebase and solve outstanding reviews, feel free to open a new PR.

**Comment by qthree** [comment]
fyi, I cherry-picked commit from this PR on top of `next` branch and added socks4/5 proxy support, feel free to use it https://github.com/qthree/serenity/tree/ws-proxy

**Comment by Morb0** [comment]
a critically important feature

**Pull request closed without merging**